<!DOCTYPE html>
<html lang="zh-CN">
    <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="robots" content="noodp" />
        <meta http-equiv="X-UA-Compatible" content="IE=edge, chrome=1">
        <title>TWAS/eQTL analysis pipeline - 美吉生物信息实验室</title><meta name="Description" content="根据基因的表达量进行eQTL分析"><meta property="og:title" content="TWAS/eQTL analysis pipeline" />
<meta property="og:description" content="根据基因的表达量进行eQTL分析" />
<meta property="og:type" content="article" />
<meta property="og:url" content="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" />
<meta property="og:image" content="http://mojorbioinfor.work/posts/matrix-eqtl-twas/eQTL.png"/>
<meta property="article:published_time" content="2020-12-14T16:29:59+08:00" />
<meta property="article:modified_time" content="2020-12-18T16:29:59+08:00" />
<meta name="twitter:card" content="summary_large_image"/>
<meta name="twitter:image" content="http://mojorbioinfor.work/posts/matrix-eqtl-twas/eQTL.png"/>
<meta name="twitter:title" content="TWAS/eQTL analysis pipeline"/>
<meta name="twitter:description" content="根据基因的表达量进行eQTL分析"/>
<meta name="application-name" content="美吉生物信息团队">
<meta name="apple-mobile-web-app-title" content="美吉生物信息团队"><meta name="theme-color" content="#ffffff"><meta name="msapplication-TileColor" content="#da532c"><link rel="icon" href="/images/MBIL.svg"><link rel="apple-touch-icon" sizes="180x180" href="/apple-touch-icon.png"><link rel="mask-icon" href="/safari-pinned-tab.svg" color="#5bbad5"><link rel="manifest" href="/site.webmanifest"><link rel="canonical" href="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" /><link rel="prev" href="http://mojorbioinfor.work/posts/probabilistic-estimation-of-expression-residuals/" /><link rel="next" href="http://mojorbioinfor.work/posts/stem-sop/" /><link rel="stylesheet" href="/lib/normalize/normalize.min.css"><link rel="stylesheet" href="/css/style.min.css"><link rel="stylesheet" href="/lib/fontawesome-free/all.min.css"><link rel="stylesheet" href="/lib/animate/animate.min.css"><script type="application/ld+json">
    {
        "@context": "http://schema.org",
        "@type": "BlogPosting",
        "headline": "TWAS/eQTL analysis pipeline",
        "inLanguage": "zh-CN",
        "mainEntityOfPage": {
            "@type": "WebPage",
            "@id": "http:\/\/mojorbioinfor.work\/posts\/matrix-eqtl-twas\/"
        },"image": [{
                            "@type": "ImageObject",
                            "url": "http:\/\/mojorbioinfor.work\/posts\/matrix-eqtl-twas\/eQTL.png",
                            "width":  1420 ,
                            "height":  588 
                        }],"genre": "posts","keywords": "eQTL, TWAS, RNAseq, 表达量, 定位","wordcount":  1661 ,
        "url": "http:\/\/mojorbioinfor.work\/posts\/matrix-eqtl-twas\/","datePublished": "2020-12-14T16:29:59+08:00","dateModified": "2020-12-18T16:29:59+08:00","publisher": {
            "@type": "Organization",
            "name": "美吉生物信息团队"},"author": {
                "@type": "Person",
                "name": "Xiaolong He"
            },"description": "根据基因的表达量进行eQTL分析"
    }
    </script></head>
    <body header-desktop="" header-mobile=""><script type="text/javascript">(window.localStorage && localStorage.getItem('theme') ? localStorage.getItem('theme') === 'dark' : ('' === 'auto' ? window.matchMedia('(prefers-color-scheme: dark)').matches : '' === 'dark')) && document.body.setAttribute('theme', 'dark');</script>

        <div id="mask"></div><div class="wrapper"><header class="desktop" id="header-desktop">
    <div class="header-wrapper">
        <div class="header-title">
            <a href="/" title="美吉生物信息实验室"><img
        class="lazyload logo"
        src="/svg/loading.min.svg"
        data-src="/images/MBIL.svg"
        data-srcset="/images/MBIL.svg, /images/MBIL.svg 1.5x, /images/MBIL.svg 2x"
        data-sizes="auto"
        alt="/images/MBIL.svg"
        title="/images/MBIL.svg" /></a>
        </div>
        <div class="menu">
            <div class="menu-inner"><a class="menu-item" href="/posts/" title="生物信息流程"> 生物信息流程 </a><a class="menu-item" href="/tags/" title="标签"> 标签 </a><a class="menu-item" href="/categories/" title="分类"> 分类 </a><a class="menu-item" href="/publication/" title="文章"> 文章 </a><a class="menu-item" href="/team/" title="团队"> 团队 </a><a class="menu-item" href="/joinus/" title="加入我们"> 加入我们 </a><a class="menu-item" href="https://github.com/MajorBio" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw'></i>  </a><span class="menu-item delimiter"></span><span class="menu-item search" id="search-desktop">
                        <input type="text" placeholder="搜索文章标题或内容..." id="search-input-desktop">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-desktop" title="搜索">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-desktop" title="清空">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-desktop">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </span><a href="javascript:void(0);" class="menu-item theme-switch" title="切换主题">
                    <i class="fas fa-adjust fa-fw"></i>
                </a>
            </div>
        </div>
    </div>
</header><header class="mobile" id="header-mobile">
    <div class="header-container">
        <div class="header-wrapper">
            <div class="header-title">
                <a href="/" title="美吉生物信息实验室"><img
        class="lazyload logo"
        src="/svg/loading.min.svg"
        data-src="/images/MBIL.svg"
        data-srcset="/images/MBIL.svg, /images/MBIL.svg 1.5x, /images/MBIL.svg 2x"
        data-sizes="auto"
        alt="/images/MBIL.svg"
        title="/images/MBIL.svg" /></a>
            </div>
            <div class="menu-toggle" id="menu-toggle-mobile">
                <span></span><span></span><span></span>
            </div>
        </div>
        <div class="menu" id="menu-mobile"><div class="search-wrapper">
                    <div class="search mobile" id="search-mobile">
                        <input type="text" placeholder="搜索文章标题或内容..." id="search-input-mobile">
                        <a href="javascript:void(0);" class="search-button search-toggle" id="search-toggle-mobile" title="搜索">
                            <i class="fas fa-search fa-fw"></i>
                        </a>
                        <a href="javascript:void(0);" class="search-button search-clear" id="search-clear-mobile" title="清空">
                            <i class="fas fa-times-circle fa-fw"></i>
                        </a>
                        <span class="search-button search-loading" id="search-loading-mobile">
                            <i class="fas fa-spinner fa-fw fa-spin"></i>
                        </span>
                    </div>
                    <a href="javascript:void(0);" class="search-cancel" id="search-cancel-mobile">
                        取消
                    </a>
                </div><a class="menu-item" href="/posts/" title="生物信息流程">生物信息流程</a><a class="menu-item" href="/tags/" title="标签">标签</a><a class="menu-item" href="/categories/" title="分类">分类</a><a class="menu-item" href="/publication/" title="文章">文章</a><a class="menu-item" href="/team/" title="团队">团队</a><a class="menu-item" href="/joinus/" title="加入我们">加入我们</a><a class="menu-item" href="https://github.com/MajorBio" title="GitHub" rel="noopener noreffer" target="_blank"><i class='fab fa-github fa-fw'></i></a><a href="javascript:void(0);" class="menu-item theme-switch" title="切换主题">
                <i class="fas fa-adjust fa-fw"></i>
            </a></div>
    </div>
</header>
<div class="search-dropdown desktop">
    <div id="search-dropdown-desktop"></div>
</div>
<div class="search-dropdown mobile">
    <div id="search-dropdown-mobile"></div>
</div>
<main class="main">
                <div class="container"><div class="toc" id="toc-auto">
            <h2 class="toc-title">目录</h2>
            <div class="toc-content" id="toc-content-auto"></div>
        </div><article class="page single"><h1 class="single-title animated flipInX">TWAS/eQTL analysis pipeline</h1><div class="post-meta">
            <div class="post-meta-line"><span class="post-author"><a href="https://github.com/xiaolong-he999" title="Author" target="_blank" rel="noopener noreffer author" class="author"><i class="fas fa-user-circle fa-fw"></i>Xiaolong He</a></span>&nbsp;<span class="post-category">收录于 <a href="/categories/rnaseq/"><i class="far fa-folder fa-fw"></i>RNAseq</a>&nbsp;<a href="/categories/pipeline/"><i class="far fa-folder fa-fw"></i>Pipeline</a></span></div>
            <div class="post-meta-line"><i class="far fa-calendar-alt fa-fw"></i>&nbsp;<time datetime="2020-12-14">2020-12-14</time>&nbsp;<i class="fas fa-pencil-alt fa-fw"></i>&nbsp;约 1661 字&nbsp;
                <i class="far fa-clock fa-fw"></i>&nbsp;预计阅读 4 分钟&nbsp;<span id="/posts/matrix-eqtl-twas/" class="leancloud_visitors" data-flag-title="TWAS/eQTL analysis pipeline">
                        <i class="far fa-eye fa-fw"></i>&nbsp;<span class=leancloud-visitors-count></span>&nbsp;次阅读
                    </span>&nbsp;</div>
        </div><div class="featured-image"><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="/posts/matrix-eqtl-twas/eQTL.png"
        data-srcset="/posts/matrix-eqtl-twas/eQTL.png, /posts/matrix-eqtl-twas/eQTL.png 1.5x, /posts/matrix-eqtl-twas/eQTL.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/eQTL.png"
        title="根据基因的表达量进行eQTL分析" /></div><div class="details toc" id="toc-static"  kept="">
                <div class="details-summary toc-title">
                    <span>目录</span>
                    <span><i class="details-icon fas fa-angle-right"></i></span>
                </div>
                <div class="details-content toc-content" id="toc-content-static"><nav id="TableOfContents">
  <ul>
    <li><a href="#数据准备">数据准备</a></li>
    <li><a href="#pipeline">Pipeline</a></li>
    <li><a href="#setp1-data-prepare">setp1 data prepare</a>
      <ul>
        <li><a href="#结果展示">结果展示：</a></li>
      </ul>
    </li>
    <li><a href="#step2-data-prepare">step2 data prepare</a>
      <ul>
        <li><a href="#结果展示-1">结果展示：</a></li>
      </ul>
    </li>
    <li><a href="#step3-calculate-eqtl-matrix">step3 calculate eQTL matrix</a>
      <ul>
        <li><a href="#结果展示-2">结果展示：</a></li>
      </ul>
    </li>
    <li><a href="#参考文献">参考文献</a></li>
  </ul>
</nav></div>
            </div><div class="content" id="content"><h1 id="eqtl">eQTL</h1>
<h2 id="数据准备">数据准备</h2>
<ol>
<li>物种的gff文件</li>
<li>vcf文件（已做完基因型填充）</li>
</ol>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">1
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-shell" data-lang="shell"><span class="nv">$java</span> -jar /mnt/ilustre/centos7users/xiaolong.he/bin/beagle/beagle.18May20.d20.jar <span class="nv">gt</span><span class="o">=</span>/mnt/ilustre/centos7users/xiaolong.he/others/GWAS_mengluo/KF/pop.uniq.vcf <span class="nv">out</span><span class="o">=</span>/mnt/ilustre/centos7users/xiaolong.he/others/GWAS_mengluo/KF/pop.phased&gt;
</code></pre></td></tr></table>
</div>
</div><ol>
<li>基因表达量矩阵（行为基因，列为样本）</li>
</ol>
<h2 id="pipeline">Pipeline</h2>
<p>示例：</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="Pipleline.png"
        data-srcset="/posts/matrix-eqtl-twas/Pipleline.png, Pipleline.png 1.5x, /posts/matrix-eqtl-twas/Pipleline.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/Pipleline.png"
        title="Pipleline.png" /></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span><span class="lnt">56
</span><span class="lnt">57
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="n">python</span> <span class="n">eQTL</span><span class="o">.</span><span class="n">Pipeline</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">e</span> <span class="n">expression</span><span class="o">.</span><span class="n">xls</span> <span class="o">-</span><span class="n">c</span> <span class="n">covariates</span> <span class="o">-</span><span class="n">g</span> <span class="n">ref</span><span class="o">.</span><span class="n">gff</span> <span class="o">-</span><span class="n">p</span> <span class="n">test</span> <span class="o">-</span><span class="n">r</span> <span class="mi">50</span> <span class="o">-</span><span class="n">o</span> <span class="o">./&gt;</span>
<span class="err">其中</span><span class="sb">`-c`</span><span class="err">参数为可选参数</span>
<span class="c1">#!/usr/bin/env python</span>
<span class="c1"># -*-coding:utf-8-*-</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>
<span class="kn">import</span> <span class="nn">time</span>

<span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&#34;GWAS.py&#34;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="s1">&#39;--vcf&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the vcf file&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-e&#39;</span><span class="p">,</span> <span class="s1">&#39;--expression&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the gene expression metrix&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-c&#39;</span><span class="p">,</span> <span class="s1">&#39;--covariates&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the covariates file&#39;</span><span class="p">,</span> <span class="n">default</span> <span class="o">=</span> <span class="s1">&#39;None&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gff&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the gff file&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-p&#39;</span><span class="p">,</span> <span class="s1">&#39;--prefix&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span><span class="n">help</span><span class="o">=</span><span class="s1">&#39;prefix out filename&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-m&#39;</span><span class="p">,</span> <span class="s1">&#39;--model&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;models, such as ANOVA/A, LINEAR/L, LINEAR_CROSS/C&#39;</span><span class="p">,</span> <span class="n">default</span> <span class="o">=</span> <span class="s1">&#39;L&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-r&#39;</span><span class="p">,</span> <span class="s1">&#39;--ram&#39;</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;RAM for running&#39;</span><span class="p">,</span> <span class="n">default</span> <span class="o">=</span> <span class="mi">20</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s1">&#39;--outpath&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the output path, which end with /&#39;</span><span class="p">)</span>
<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">data_prepare</span><span class="p">():</span>
    <span class="n">out</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">outpath</span><span class="p">)</span>
    <span class="n">vcf</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">vcf</span><span class="p">)</span>
    <span class="n">gff</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">)</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="n">out</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;work_sh&#39;</span><span class="p">):</span>
        <span class="k">pass</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="s1">&#39;work_sh&#39;</span><span class="p">)</span>
    <span class="n">scpath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>
    <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="s2">&#34;work_sh/eqtl.sh&#34;</span><span class="p">,</span> <span class="s2">&#34;w&#34;</span><span class="p">)</span>
    <span class="n">step1</span> <span class="o">=</span> <span class="n">scpath</span><span class="o">+</span><span class="s1">&#39;/step01.vcf2num.R&#39;</span>
    <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;cd &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="s2">&#34; &amp;&amp; &#34;</span><span class="p">)</span>
    <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;Rscript &#34;</span> <span class="o">+</span> <span class="n">step1</span> <span class="o">+</span> <span class="s2">&#34; -v &#34;</span> <span class="o">+</span> <span class="n">vcf</span> <span class="o">+</span> <span class="s2">&#34; -o &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="s2">&#34; -p &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&#34; &amp; &#34;</span><span class="p">)</span>
    <span class="c1"># f2 = open(&#34;work_sh/step02.gff2geneloc.sh&#34;, &#34;w&#34;)</span>
    <span class="n">step2</span> <span class="o">=</span> <span class="n">scpath</span><span class="o">+</span><span class="s1">&#39;/step02.gff2geneloc.py&#39;</span>
    <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;python &#34;</span> <span class="o">+</span> <span class="n">step2</span> <span class="o">+</span> <span class="s2">&#34; -g &#34;</span> <span class="o">+</span> <span class="n">gff</span> <span class="o">+</span> <span class="s2">&#34; -o &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="s1">&#39; &amp;&amp; &#39;</span><span class="p">)</span>
    <span class="c1"># f3 = open(&#34;work_sh/step03.eqtl.sh&#34;, &#34;w&#34;)</span>
    <span class="n">step3</span> <span class="o">=</span> <span class="n">scpath</span><span class="o">+</span><span class="s1">&#39;/step03.eqtl.R&#39;</span>
    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">covariates</span> <span class="o">!=</span> <span class="s2">&#34;None&#34;</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;export R_LIBS=;export R_HOME=;export RHOME=;/mnt/ilustre/centos7users/dna/.env/R-3.6.1/bin/Rscript &#34;</span><span class="o">+</span> <span class="n">step3</span> \
                <span class="o">+</span> <span class="s2">&#34; -s &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&#34;.SNP.txt&#34;</span> <span class="o">+</span> <span class="s2">&#34; -e &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">expression</span> <span class="o">+</span> <span class="s2">&#34; -c &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">covariates</span> <span class="o">+</span> <span class="s2">&#34; -l &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> \
                <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&#34;.snpsloc.txt&#34;</span> <span class="o">+</span> <span class="s2">&#34; -g &#34;</span><span class="o">+</span> <span class="s2">&#34;geneloc.txt&#34;</span> <span class="o">+</span> <span class="s2">&#34; -o &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">outpath</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">f</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;export R_LIBS=;export R_HOME=;export RHOME=;/mnt/ilustre/centos7users/dna/.env/R-3.6.1/bin/Rscript &#34;</span> <span class="o">+</span> <span class="n">step3</span> \
                <span class="o">+</span> <span class="s2">&#34; -s &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&#34;.SNP.txt&#34;</span> <span class="o">+</span> <span class="s2">&#34; -e &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">expression</span> <span class="o">+</span> <span class="s2">&#34; -l &#34;</span> <span class="o">+</span> <span class="n">out</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">prefix</span> <span class="o">+</span> <span class="s2">&#34;.snpsloc.txt&#34;</span> \
                <span class="o">+</span> <span class="s2">&#34; -g &#34;</span><span class="o">+</span> <span class="s2">&#34;geneloc.txt&#34;</span> <span class="o">+</span> <span class="s2">&#34; -m &#34;</span> <span class="o">+</span> <span class="n">args</span><span class="o">.</span><span class="n">model</span> <span class="o">+</span> <span class="s2">&#34; -o &#34;</span> <span class="o">+</span> <span class="n">out</span><span class="p">)</span>
    <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="s2">&#34;work_sh&#34;</span><span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">Popen</span><span class="p">([</span><span class="s2">&#34;qsub-slurm.pl&#34;</span><span class="p">,</span> <span class="s2">&#34;--Resource&#34;</span><span class="p">,</span> <span class="s2">&#34;mem=&#34;</span><span class="o">+</span><span class="n">args</span><span class="o">.</span><span class="n">ram</span><span class="p">,</span> <span class="s2">&#34;--CPU&#34;</span><span class="p">,</span> <span class="s2">&#34;2&#34;</span><span class="p">,</span> <span class="s2">&#34;eqtl.sh&#34;</span><span class="p">])</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">data_prepare</span><span class="p">()</span>
    <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
    <span class="k">print</span><span class="p">(</span><span class="s1">&#39;Running time: </span><span class="si">%s</span><span class="s1"> Seconds&#39;</span><span class="o">%</span><span class="p">(</span><span class="n">end</span><span class="o">-</span><span class="n">start</span><span class="p">))</span>
</code></pre></td></tr></table>
</div>
</div><h2 id="setp1-data-prepare">setp1 data prepare</h2>
<p>从vcf文件提取snp位置信息，以及snp基因型矩阵</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step01.vcf2num.png"
        data-srcset="/posts/matrix-eqtl-twas/step01.vcf2num.png, step01.vcf2num.png 1.5x, /posts/matrix-eqtl-twas/step01.vcf2num.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step01.vcf2num.png"
        title="step01.vcf2num.png" /></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span><span class="lnt">38
</span><span class="lnt">39
</span><span class="lnt">40
</span><span class="lnt">41
</span><span class="lnt">42
</span><span class="lnt">43
</span><span class="lnt">44
</span><span class="lnt">45
</span><span class="lnt">46
</span><span class="lnt">47
</span><span class="lnt">48
</span><span class="lnt">49
</span><span class="lnt">50
</span><span class="lnt">51
</span><span class="lnt">52
</span><span class="lnt">53
</span><span class="lnt">54
</span><span class="lnt">55
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#/usr/bin/env Rscript</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;/mnt/ilustre/centos7users/meng.luo/.env/lib/R/library/&#34;</span><span class="p">,</span><span class="s">&#34;/mnt/ilustre/centos7users/xiaolong.he/lib/R3.5/&#34;</span><span class="p">))</span>

<span class="nf">library</span><span class="p">(</span><span class="s">&#39;getopt&#39;</span><span class="p">)</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
		<span class="s">&#39;vcf&#39;</span><span class="p">,</span><span class="s">&#39;v&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
		<span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
		<span class="s">&#39;prefix&#39;</span><span class="p">,</span><span class="s">&#39;p&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
		<span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
		<span class="p">),</span><span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
	<span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">))</span>
	<span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">)</span>
	<span class="nf">cat</span><span class="p">(</span><span class="s">&#34;
</span><span class="s">Usage example:
</span><span class="s">	Rscript vcf2num.R
</span><span class="s">Usage:
</span><span class="s">	--vcf vcf
</span><span class="s">	--outdir outdir
</span><span class="s">	--prefix prefix name for out
</span><span class="s">	--help usage\n&#34;</span><span class="p">)</span>
	<span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">vcf</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>

<span class="nf">setwd</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">rMVP</span><span class="p">)</span>
<span class="nf">MVP.Data.VCF2MVP</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">vcf</span><span class="p">,</span> <span class="n">out</span><span class="o">=</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
<span class="n">geno</span> <span class="o">&lt;-</span> <span class="nf">attach.big.matrix</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#34;geno.desc&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;.&#39;</span><span class="p">))</span>
<span class="n">map</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#34;geno.map&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;.&#39;</span><span class="p">),</span><span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">stringsAsFactors</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">samples</span> <span class="o">&lt;-</span> <span class="nf">read.table</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#34;geno.ind&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;.&#39;</span><span class="p">),</span><span class="n">header</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">stringsAsFactors</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">dat</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">geno[</span><span class="p">,</span><span class="n">]</span><span class="p">)</span>
<span class="c1"># dat[is.na(dat)] &lt;- 0</span>
<span class="c1"># new_dat &lt;- cbind(map$SNP,dat)</span>
<span class="n">dat_imu</span> <span class="o">&lt;-</span> <span class="nf">matrix</span><span class="p">(</span><span class="n">nrow</span><span class="o">=</span><span class="nf">dim</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span><span class="n">[1]</span><span class="p">,</span><span class="n">ncol</span><span class="o">=</span><span class="nf">dim</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span><span class="n">[2]</span><span class="p">)</span>
<span class="nf">for</span><span class="p">(</span><span class="n">i</span> <span class="n">in</span> <span class="m">1</span><span class="o">:</span><span class="nf">dim</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span><span class="n">[1]</span><span class="p">){</span>
	<span class="n">line</span> <span class="o">&lt;-</span> <span class="nf">as.numeric</span><span class="p">(</span><span class="n">dat[i</span><span class="p">,</span><span class="n">]</span><span class="p">)</span>
	<span class="n">aver</span> <span class="o">&lt;-</span> <span class="nf">mean</span><span class="p">(</span><span class="n">line</span><span class="p">,</span><span class="n">na.rm</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
	<span class="n">line</span><span class="nf">[is.na</span><span class="p">(</span><span class="n">line</span><span class="p">)</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="n">aver</span>
	<span class="n">dat_imu[i</span><span class="p">,</span><span class="n">]</span> <span class="o">&lt;-</span> <span class="n">line</span>
<span class="p">}</span>
<span class="n">new_dat</span> <span class="o">&lt;-</span> <span class="nf">cbind</span><span class="p">(</span><span class="n">map</span><span class="o">$</span><span class="n">SNP</span><span class="p">,</span><span class="n">dat_imu</span><span class="p">)</span>
<span class="nf">colnames</span><span class="p">(</span><span class="n">new_dat</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&#34;snpid&#34;</span><span class="p">,</span><span class="n">samples[</span><span class="p">,</span><span class="m">1</span><span class="n">]</span><span class="p">)</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">new_dat</span><span class="p">,</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#39;SNP.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;.&#39;</span><span class="p">),</span><span class="n">col.names</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="n">snpsloc</span> <span class="o">&lt;-</span> <span class="n">map[</span><span class="p">,</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="n">]</span><span class="p">;</span><span class="nf">names</span><span class="p">(</span><span class="n">snpsloc</span><span class="p">)</span><span class="o">&lt;-</span><span class="nf">c</span><span class="p">(</span><span class="s">&#34;snpid&#34;</span><span class="p">,</span><span class="s">&#34;chr&#34;</span><span class="p">,</span><span class="s">&#34;pos&#34;</span><span class="p">)</span>
<span class="nf">write.table</span><span class="p">(</span><span class="n">snpsloc</span><span class="p">,</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#39;snpsloc.txt&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;.&#39;</span><span class="p">),</span><span class="n">col.names</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span><span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;\t&#39;</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">prefix</span><span class="p">,</span><span class="s">&#34;RData&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#34;.&#34;</span><span class="p">))</span>
<span class="nf">print</span><span class="p">(</span><span class="s">&#34;This-work-has-been-completed!&#34;</span><span class="p">)</span>
<span class="nf">proc.time</span><span class="p">()</span>
</code></pre></td></tr></table>
</div>
</div><h3 id="结果展示">结果展示：</h3>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step01.result.png"
        data-srcset="/posts/matrix-eqtl-twas/step01.result.png, step01.result.png 1.5x, /posts/matrix-eqtl-twas/step01.result.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step01.result.png"
        title="step01.result.png" /></p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step01.result-details.png"
        data-srcset="/posts/matrix-eqtl-twas/step01.result-details.png, step01.result-details.png 1.5x, /posts/matrix-eqtl-twas/step01.result-details.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step01.result-details.png"
        title="step01.result-details.png" /></p>
<h2 id="step2-data-prepare">step2 data prepare</h2>
<p>从gff文件提取gene位置信息</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step02.gff2geneloc.png"
        data-srcset="/posts/matrix-eqtl-twas/step02.gff2geneloc.png, step02.gff2geneloc.png 1.5x, /posts/matrix-eqtl-twas/step02.gff2geneloc.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step02.gff2geneloc.png"
        title="step02.gff2geneloc.png" /></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt"> 1
</span><span class="lnt"> 2
</span><span class="lnt"> 3
</span><span class="lnt"> 4
</span><span class="lnt"> 5
</span><span class="lnt"> 6
</span><span class="lnt"> 7
</span><span class="lnt"> 8
</span><span class="lnt"> 9
</span><span class="lnt">10
</span><span class="lnt">11
</span><span class="lnt">12
</span><span class="lnt">13
</span><span class="lnt">14
</span><span class="lnt">15
</span><span class="lnt">16
</span><span class="lnt">17
</span><span class="lnt">18
</span><span class="lnt">19
</span><span class="lnt">20
</span><span class="lnt">21
</span><span class="lnt">22
</span><span class="lnt">23
</span><span class="lnt">24
</span><span class="lnt">25
</span><span class="lnt">26
</span><span class="lnt">27
</span><span class="lnt">28
</span><span class="lnt">29
</span><span class="lnt">30
</span><span class="lnt">31
</span><span class="lnt">32
</span><span class="lnt">33
</span><span class="lnt">34
</span><span class="lnt">35
</span><span class="lnt">36
</span><span class="lnt">37
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-python" data-lang="python"><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*-coding:utf-8-*-</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">subprocess</span>

<span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span><span class="n">description</span><span class="o">=</span><span class="s2">&#34;GWAS.py&#34;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-g&#39;</span><span class="p">,</span> <span class="s1">&#39;--gff&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the gff file&#39;</span><span class="p">)</span>
<span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-o&#39;</span><span class="p">,</span> <span class="s1">&#39;--outpath&#39;</span><span class="p">,</span> <span class="n">required</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s1">&#39;the output path, which end with /&#39;</span><span class="p">)</span>
<span class="n">args</span> <span class="o">=</span> <span class="n">parser</span><span class="o">.</span><span class="n">parse_args</span><span class="p">()</span>
<span class="k">def</span> <span class="nf">gff2geneloc</span><span class="p">():</span>
    <span class="n">f</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">,</span> <span class="s1">&#39;r&#39;</span><span class="p">)</span>
    <span class="n">fo</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">outpath</span> <span class="o">+</span> <span class="s2">&#34;geneloc.txt&#34;</span><span class="p">,</span> <span class="s2">&#34;w&#34;</span><span class="p">)</span>
    <span class="n">fo</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="s2">&#34;geneid</span><span class="se">\t</span><span class="s2">chr</span><span class="se">\t</span><span class="s2">left</span><span class="se">\t</span><span class="s2">right</span><span class="se">\n</span><span class="s2">&#34;</span><span class="p">)</span>
    <span class="n">geneloc</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">genes</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">f</span><span class="p">:</span>
        <span class="n">Chr</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> 
        <span class="n">loc1</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span>
        <span class="n">loc2</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">4</span><span class="p">]</span>
        <span class="n">geneid</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;=&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">loc1</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">int</span><span class="p">(</span><span class="n">loc2</span><span class="p">):</span>
            <span class="n">left</span> <span class="o">=</span> <span class="n">loc1</span>
            <span class="n">right</span> <span class="o">=</span> <span class="n">loc2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">left</span> <span class="o">=</span> <span class="n">loc2</span>
            <span class="n">right</span> <span class="o">=</span> <span class="n">loc1</span>
        <span class="k">if</span> <span class="n">geneid</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">genes</span><span class="p">:</span>
            <span class="n">genes</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneid</span><span class="p">)</span>
            <span class="n">geneloc</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">geneid</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">Chr</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">left</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="o">+</span> <span class="n">right</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">fo</span><span class="o">.</span><span class="n">writelines</span><span class="p">(</span><span class="n">i</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">geneloc</span><span class="p">)</span>
    <span class="n">f</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="n">fo</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>

<span class="k">if</span> <span class="vm">__name__</span> <span class="o">==</span> <span class="s1">&#39;__main__&#39;</span><span class="p">:</span>
    <span class="n">gff2geneloc</span><span class="p">()</span>
</code></pre></td></tr></table>
</div>
</div><h3 id="结果展示-1">结果展示：</h3>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step02.result-details.png"
        data-srcset="/posts/matrix-eqtl-twas/step02.result-details.png, step02.result-details.png 1.5x, /posts/matrix-eqtl-twas/step02.result-details.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step02.result-details.png"
        title="step02.result-details.png" /></p>
<h2 id="step3-calculate-eqtl-matrix">step3 calculate eQTL matrix</h2>
<p>利用基因表达矩阵、基因位置文件、snp、snp位置文件、协变量矩阵（可选），计算cis、trans作用位点</p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step03.eqtl.png"
        data-srcset="/posts/matrix-eqtl-twas/step03.eqtl.png, step03.eqtl.png 1.5x, /posts/matrix-eqtl-twas/step03.eqtl.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step03.eqtl.png"
        title="step03.eqtl.png" /></p>
<div class="highlight"><div class="chroma">
<table class="lntable"><tr><td class="lntd">
<pre class="chroma"><code><span class="lnt">  1
</span><span class="lnt">  2
</span><span class="lnt">  3
</span><span class="lnt">  4
</span><span class="lnt">  5
</span><span class="lnt">  6
</span><span class="lnt">  7
</span><span class="lnt">  8
</span><span class="lnt">  9
</span><span class="lnt"> 10
</span><span class="lnt"> 11
</span><span class="lnt"> 12
</span><span class="lnt"> 13
</span><span class="lnt"> 14
</span><span class="lnt"> 15
</span><span class="lnt"> 16
</span><span class="lnt"> 17
</span><span class="lnt"> 18
</span><span class="lnt"> 19
</span><span class="lnt"> 20
</span><span class="lnt"> 21
</span><span class="lnt"> 22
</span><span class="lnt"> 23
</span><span class="lnt"> 24
</span><span class="lnt"> 25
</span><span class="lnt"> 26
</span><span class="lnt"> 27
</span><span class="lnt"> 28
</span><span class="lnt"> 29
</span><span class="lnt"> 30
</span><span class="lnt"> 31
</span><span class="lnt"> 32
</span><span class="lnt"> 33
</span><span class="lnt"> 34
</span><span class="lnt"> 35
</span><span class="lnt"> 36
</span><span class="lnt"> 37
</span><span class="lnt"> 38
</span><span class="lnt"> 39
</span><span class="lnt"> 40
</span><span class="lnt"> 41
</span><span class="lnt"> 42
</span><span class="lnt"> 43
</span><span class="lnt"> 44
</span><span class="lnt"> 45
</span><span class="lnt"> 46
</span><span class="lnt"> 47
</span><span class="lnt"> 48
</span><span class="lnt"> 49
</span><span class="lnt"> 50
</span><span class="lnt"> 51
</span><span class="lnt"> 52
</span><span class="lnt"> 53
</span><span class="lnt"> 54
</span><span class="lnt"> 55
</span><span class="lnt"> 56
</span><span class="lnt"> 57
</span><span class="lnt"> 58
</span><span class="lnt"> 59
</span><span class="lnt"> 60
</span><span class="lnt"> 61
</span><span class="lnt"> 62
</span><span class="lnt"> 63
</span><span class="lnt"> 64
</span><span class="lnt"> 65
</span><span class="lnt"> 66
</span><span class="lnt"> 67
</span><span class="lnt"> 68
</span><span class="lnt"> 69
</span><span class="lnt"> 70
</span><span class="lnt"> 71
</span><span class="lnt"> 72
</span><span class="lnt"> 73
</span><span class="lnt"> 74
</span><span class="lnt"> 75
</span><span class="lnt"> 76
</span><span class="lnt"> 77
</span><span class="lnt"> 78
</span><span class="lnt"> 79
</span><span class="lnt"> 80
</span><span class="lnt"> 81
</span><span class="lnt"> 82
</span><span class="lnt"> 83
</span><span class="lnt"> 84
</span><span class="lnt"> 85
</span><span class="lnt"> 86
</span><span class="lnt"> 87
</span><span class="lnt"> 88
</span><span class="lnt"> 89
</span><span class="lnt"> 90
</span><span class="lnt"> 91
</span><span class="lnt"> 92
</span><span class="lnt"> 93
</span><span class="lnt"> 94
</span><span class="lnt"> 95
</span><span class="lnt"> 96
</span><span class="lnt"> 97
</span><span class="lnt"> 98
</span><span class="lnt"> 99
</span><span class="lnt">100
</span><span class="lnt">101
</span><span class="lnt">102
</span><span class="lnt">103
</span><span class="lnt">104
</span><span class="lnt">105
</span><span class="lnt">106
</span><span class="lnt">107
</span><span class="lnt">108
</span><span class="lnt">109
</span><span class="lnt">110
</span><span class="lnt">111
</span><span class="lnt">112
</span><span class="lnt">113
</span><span class="lnt">114
</span><span class="lnt">115
</span><span class="lnt">116
</span><span class="lnt">117
</span><span class="lnt">118
</span><span class="lnt">119
</span><span class="lnt">120
</span><span class="lnt">121
</span><span class="lnt">122
</span><span class="lnt">123
</span><span class="lnt">124
</span><span class="lnt">125
</span><span class="lnt">126
</span><span class="lnt">127
</span><span class="lnt">128
</span><span class="lnt">129
</span><span class="lnt">130
</span><span class="lnt">131
</span><span class="lnt">132
</span><span class="lnt">133
</span><span class="lnt">134
</span><span class="lnt">135
</span><span class="lnt">136
</span><span class="lnt">137
</span><span class="lnt">138
</span><span class="lnt">139
</span><span class="lnt">140
</span><span class="lnt">141
</span><span class="lnt">142
</span><span class="lnt">143
</span><span class="lnt">144
</span><span class="lnt">145
</span><span class="lnt">146
</span><span class="lnt">147
</span><span class="lnt">148
</span><span class="lnt">149
</span><span class="lnt">150
</span><span class="lnt">151
</span><span class="lnt">152
</span><span class="lnt">153
</span><span class="lnt">154
</span><span class="lnt">155
</span><span class="lnt">156
</span><span class="lnt">157
</span><span class="lnt">158
</span><span class="lnt">159
</span><span class="lnt">160
</span><span class="lnt">161
</span><span class="lnt">162
</span><span class="lnt">163
</span><span class="lnt">164
</span><span class="lnt">165
</span><span class="lnt">166
</span><span class="lnt">167
</span><span class="lnt">168
</span><span class="lnt">169
</span><span class="lnt">170
</span><span class="lnt">171
</span><span class="lnt">172
</span><span class="lnt">173
</span><span class="lnt">174
</span><span class="lnt">175
</span><span class="lnt">176
</span><span class="lnt">177
</span><span class="lnt">178
</span></code></pre></td>
<td class="lntd">
<pre class="chroma"><code class="language-r" data-lang="r"><span class="c1">#!/usr/bin/env Rscript3.6</span>
<span class="nf">.libPaths</span><span class="p">(</span><span class="s">&#34;/mnt/ilustre/centos7users/xiaolong.he/R/x86_64-pc-linux-gnu-library/3.6/&#34;</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="s">&#39;getopt&#39;</span><span class="p">);</span>
<span class="nf">options</span><span class="p">(</span><span class="n">bitmapType</span><span class="o">=</span><span class="s">&#39;cairo&#39;</span><span class="p">)</span>
<span class="n">spec</span> <span class="o">=</span> <span class="nf">matrix</span><span class="p">(</span><span class="nf">c</span><span class="p">(</span>
	<span class="s">&#39;snps&#39;</span><span class="p">,</span><span class="s">&#39;s&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;expressions&#39;</span><span class="p">,</span><span class="s">&#39;e&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;covariates&#39;</span><span class="p">,</span><span class="s">&#39;c&#39;</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;snpspos&#39;</span><span class="p">,</span><span class="s">&#39;l&#39;</span><span class="p">,</span> <span class="m">2</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;genespos&#39;</span><span class="p">,</span><span class="s">&#39;g&#39;</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;outdir&#39;</span><span class="p">,</span><span class="s">&#39;o&#39;</span><span class="p">,</span><span class="m">1</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;threshold&#39;</span><span class="p">,</span><span class="s">&#39;t&#39;</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="s">&#39;numeric&#39;</span><span class="p">,</span>
	<span class="s">&#39;distance&#39;</span><span class="p">,</span><span class="s">&#39;d&#39;</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="s">&#39;numeric&#39;</span><span class="p">,</span>
	<span class="s">&#39;model&#39;</span><span class="p">,</span><span class="s">&#39;m&#39;</span><span class="p">,</span><span class="m">2</span><span class="p">,</span><span class="s">&#39;character&#39;</span><span class="p">,</span>
	<span class="s">&#39;help&#39;</span><span class="p">,</span><span class="s">&#39;h&#39;</span><span class="p">,</span><span class="m">0</span><span class="p">,</span><span class="s">&#39;logical&#39;</span>
	<span class="p">),</span> <span class="n">byrow</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">ncol</span><span class="o">=</span><span class="m">4</span><span class="p">);</span>
<span class="n">opt</span> <span class="o">=</span> <span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">);</span>
<span class="n">print_usage</span> <span class="o">&lt;-</span> <span class="nf">function</span><span class="p">(</span><span class="n">spec</span><span class="o">=</span><span class="kc">NULL</span><span class="p">){</span>
	<span class="nf">cat</span><span class="p">(</span><span class="nf">getopt</span><span class="p">(</span><span class="n">spec</span><span class="p">,</span> <span class="n">usage</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">));</span>
	<span class="nf">cat</span><span class="p">(</span><span class="s">&#34;Usage example: \n&#34;</span><span class="p">);</span>
	<span class="nf">cat</span><span class="p">(</span><span class="s">&#34;
</span><span class="s">Usage example:
</span><span class="s">	Rscript eqtl.R
</span><span class="s">
</span><span class="s">Usage:
</span><span class="s">	--snps	the snps file
</span><span class="s">	--expressions	the expressions file
</span><span class="s">	--covariates	the covariates file
</span><span class="s">	--snpspos	snps position file
</span><span class="s">	--genespos	genes position file
</span><span class="s">	--threshold	threshold, defaule 1e-2
</span><span class="s">	--distance	distance, default 1e6
</span><span class="s">	--model	model, such as ANOVA/A, LINEAR/L, LINEAR_CROSS/C, default LINEAR
</span><span class="s">	--outdir	outdir
</span><span class="s">	--help	usage\n&#34;</span>
	<span class="p">);</span>
	<span class="nf">q</span><span class="p">(</span><span class="n">status</span><span class="o">=</span><span class="m">1</span><span class="p">);</span>
<span class="p">}</span>

<span class="nf">if </span><span class="p">(</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">help</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">snps</span><span class="p">))</span> <span class="p">{</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">expressions</span><span class="p">)){</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">threshold</span><span class="p">)){</span> <span class="n">opt</span><span class="o">$</span><span class="n">threshold</span> <span class="o">=</span> <span class="m">1e-2</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">distance</span><span class="p">)){</span> <span class="n">opt</span><span class="o">$</span><span class="n">distance</span> <span class="o">=</span> <span class="m">1e6</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">model</span><span class="p">)){</span> <span class="n">opt</span><span class="o">$</span><span class="n">model</span> <span class="o">=</span> <span class="s">&#39;A&#39;</span> <span class="p">}</span>
<span class="nf">if </span><span class="p">(</span> <span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">)){</span> <span class="nf">print_usage</span><span class="p">(</span><span class="n">spec</span><span class="p">)</span> <span class="p">}</span>
<span class="c1"># if ( is.null(opt$covariates)){ print_usage(spec) }</span>


<span class="nf">suppressMessages</span><span class="p">(</span><span class="nf">library</span><span class="p">(</span><span class="n">MatrixEQTL</span><span class="p">))</span>

<span class="nf">if</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">model</span> <span class="o">==</span> <span class="s">&#39;L&#39;</span><span class="p">){</span>
	<span class="n">useModel</span> <span class="o">=</span> <span class="n">modelLINEAR</span><span class="p">;</span>
<span class="p">}</span><span class="n">else</span> <span class="nf">if</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">model</span> <span class="o">==</span> <span class="s">&#39;A&#39;</span><span class="p">){</span>
	<span class="n">useModel</span> <span class="o">=</span> <span class="n">modelANOVA</span>
<span class="p">}</span><span class="n">else</span> <span class="nf">if</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">model</span> <span class="o">==</span> <span class="s">&#39;C&#39;</span><span class="p">){</span>
	<span class="n">useModel</span> <span class="o">=</span> <span class="n">modelLINEAR_CROSS</span>
<span class="p">}</span>

<span class="n">snps</span> <span class="o">=</span> <span class="n">SlicedData</span><span class="o">$</span><span class="nf">new</span><span class="p">();</span>
<span class="n">snps</span><span class="o">$</span><span class="n">fileDelimiter</span> <span class="o">=</span> <span class="s">&#34;\t&#34;</span><span class="p">;</span>      <span class="c1"># the TAB character</span>
<span class="n">snps</span><span class="o">$</span><span class="n">fileOmitCharacters</span> <span class="o">=</span> <span class="s">&#34;NA&#34;</span><span class="p">;</span> <span class="c1"># denote missing values;</span>
<span class="n">snps</span><span class="o">$</span><span class="n">fileSkipRows</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>          <span class="c1"># one row of column labels</span>
<span class="n">snps</span><span class="o">$</span><span class="n">fileSkipColumns</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>       <span class="c1"># one column of row labels</span>
<span class="n">snps</span><span class="o">$</span><span class="nf">LoadFile</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">snps</span><span class="p">);</span>

<span class="n">gene</span> <span class="o">=</span> <span class="n">SlicedData</span><span class="o">$</span><span class="nf">new</span><span class="p">();</span>
<span class="n">gene</span><span class="o">$</span><span class="n">fileDelimiter</span> <span class="o">=</span> <span class="s">&#34;\t&#34;</span><span class="p">;</span>      <span class="c1"># the TAB character</span>
<span class="n">gene</span><span class="o">$</span><span class="n">fileOmitCharacters</span> <span class="o">=</span> <span class="s">&#34;NA&#34;</span><span class="p">;</span> <span class="c1"># denote missing values;</span>
<span class="n">gene</span><span class="o">$</span><span class="n">fileSkipRows</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>          <span class="c1"># one row of column labels</span>
<span class="n">gene</span><span class="o">$</span><span class="n">fileSkipColumns</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>       <span class="c1"># one column of row labels</span>
<span class="n">gene</span><span class="o">$</span><span class="nf">LoadFile</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">expressions</span><span class="p">);</span>
<span class="c1"># errorCovariance = numeric();</span>
<span class="n">pvOutputThreshold</span> <span class="o">=</span> <span class="n">opt</span><span class="o">$</span><span class="n">threshold</span>
<span class="n">cvrt</span> <span class="o">=</span> <span class="n">SlicedData</span><span class="o">$</span><span class="nf">new</span><span class="p">();</span>
<span class="n">cvrt</span><span class="o">$</span><span class="n">fileDelimiter</span> <span class="o">=</span> <span class="s">&#34;\t&#34;</span><span class="p">;</span>      <span class="c1"># the TAB character</span>
<span class="n">cvrt</span><span class="o">$</span><span class="n">fileOmitCharacters</span> <span class="o">=</span> <span class="s">&#34;NA&#34;</span><span class="p">;</span> <span class="c1"># denote missing values;</span>
<span class="n">cvrt</span><span class="o">$</span><span class="n">fileSkipRows</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>          <span class="c1"># one row of column labels</span>
<span class="n">cvrt</span><span class="o">$</span><span class="n">fileSkipColumns</span> <span class="o">=</span> <span class="m">1</span><span class="p">;</span>       <span class="c1"># one column of row labels</span>
<span class="nf">if</span><span class="p">(</span><span class="nf">length</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">covariates</span><span class="p">)</span><span class="o">&gt;</span><span class="m">0</span><span class="p">)</span> <span class="p">{</span>
	<span class="n">cvrt</span><span class="o">$</span><span class="nf">LoadFile</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">covariates</span><span class="p">);</span>
<span class="p">}</span>


<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">snpspos</span><span class="p">)</span> <span class="o">&amp;</span> <span class="o">!</span><span class="nf">is.null</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">genespos</span><span class="p">)){</span>
<span class="n">output_file_name_cis</span> <span class="o">=</span> <span class="nf">tempfile</span><span class="p">();</span>
<span class="n">output_file_name_tra</span> <span class="o">=</span> <span class="nf">tempfile</span><span class="p">();</span>
<span class="n">pvOutputThreshold_cis</span> <span class="o">=</span> <span class="n">opt</span><span class="o">$</span><span class="n">threshold</span><span class="p">;</span>
<span class="n">pvOutputThreshold_tra</span> <span class="o">=</span> <span class="n">opt</span><span class="o">$</span><span class="n">threshold</span><span class="p">;</span>
<span class="n">cisDist</span> <span class="o">=</span> <span class="n">opt</span><span class="o">$</span><span class="n">distance</span><span class="p">;</span>
<span class="n">snpspos</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">snpspos</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">);</span>
<span class="n">genepos</span> <span class="o">=</span> <span class="nf">read.table</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">genespos</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">stringsAsFactors</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">);</span>
<span class="n">meq</span> <span class="o">=</span> <span class="nf">Matrix_eQTL_main</span><span class="p">(</span>
	<span class="n">snps</span> <span class="o">=</span> <span class="n">snps</span><span class="p">,</span>
	<span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
	<span class="n">cvrt</span> <span class="o">=</span> <span class="n">cvrt</span><span class="p">,</span>
	<span class="n">output_file_name</span> <span class="o">=</span> <span class="n">output_file_name_tra</span><span class="p">,</span>
	<span class="n">pvOutputThreshold</span> <span class="o">=</span> <span class="n">pvOutputThreshold_tra</span><span class="p">,</span>
	<span class="n">useModel</span> <span class="o">=</span> <span class="n">useModel</span><span class="p">,</span>
	<span class="c1"># errorCovariance = errorCovariance,</span>
	<span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
	<span class="n">output_file_name.cis</span> <span class="o">=</span> <span class="n">output_file_name_cis</span><span class="p">,</span>
	<span class="n">pvOutputThreshold.cis</span> <span class="o">=</span> <span class="n">pvOutputThreshold_cis</span><span class="p">,</span>
	<span class="n">snpspos</span> <span class="o">=</span> <span class="n">snpspos</span><span class="p">,</span>
	<span class="n">genepos</span> <span class="o">=</span> <span class="n">genepos</span><span class="p">,</span>
	<span class="n">cisDist</span> <span class="o">=</span> <span class="n">cisDist</span><span class="p">,</span>
	<span class="n">pvalue.hist</span> <span class="o">=</span> <span class="s">&#34;qqplot&#34;</span><span class="p">,</span>
	<span class="n">min.pv.by.genesnp</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
	<span class="n">noFDRsaveMemory</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">);</span>
<span class="nf">unlink</span><span class="p">(</span><span class="n">output_file_name_tra</span><span class="p">);</span>
<span class="nf">unlink</span><span class="p">(</span><span class="n">output_file_name_cis</span><span class="p">);</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">meq</span><span class="o">$</span><span class="n">cis</span><span class="o">$</span><span class="n">eqtl</span><span class="p">,</span> <span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#39;cis.eqtl.csv&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">),</span><span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">write.csv</span><span class="p">(</span><span class="n">meq</span><span class="o">$</span><span class="n">trans</span><span class="o">$</span><span class="n">eqtl</span><span class="p">,</span> <span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#39;trans.eqtl.csv&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">),</span><span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="nf">save.image</span><span class="p">(</span><span class="s">&#34;meq.RData&#34;</span><span class="p">)</span>
<span class="n">meh</span> <span class="o">=</span> <span class="nf">Matrix_eQTL_main</span><span class="p">(</span>
	<span class="n">snps</span> <span class="o">=</span> <span class="n">snps</span><span class="p">,</span>
        <span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
        <span class="n">cvrt</span> <span class="o">=</span> <span class="n">cvrt</span><span class="p">,</span>
        <span class="n">output_file_name</span> <span class="o">=</span> <span class="n">output_file_name_tra</span><span class="p">,</span>
        <span class="n">pvOutputThreshold</span> <span class="o">=</span> <span class="n">pvOutputThreshold_tra</span><span class="p">,</span>
	<span class="n">useModel</span> <span class="o">=</span> <span class="n">useModel</span><span class="p">,</span>
        <span class="c1"># errorCovariance = numeric(),</span>
	<span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
	<span class="n">output_file_name.cis</span> <span class="o">=</span> <span class="n">output_file_name_cis</span><span class="p">,</span>
	<span class="n">pvOutputThreshold.cis</span> <span class="o">=</span> <span class="n">pvOutputThreshold_cis</span><span class="p">,</span>
	<span class="n">snpspos</span> <span class="o">=</span> <span class="n">snpspos</span><span class="p">,</span>
	<span class="n">genepos</span> <span class="o">=</span> <span class="n">genepos</span><span class="p">,</span><span class="n">cisDist</span> <span class="o">=</span> <span class="n">cisDist</span><span class="p">,</span>
	<span class="n">pvalue.hist</span> <span class="o">=</span> <span class="m">100</span><span class="p">,</span>
	<span class="n">min.pv.by.genesnp</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span>
	<span class="n">noFDRsaveMemory</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">);</span>
<span class="c1">## Results:</span>
<span class="c1"># write.csv(meq$cis$eqtl, paste(opt$outdir,&#39;cis.eqtl.csv&#39;,sep=&#39;&#39;),quote=F)</span>
<span class="c1"># write.csv(meq$trans$eqtl, paste(opt$outdir,&#39;trans.eqtl.csv&#39;,sep=&#39;&#39;),quote=F)</span>
<span class="c1"># cat(&#39;Analysis done in: &#39;, meq$time.in.sec+meh$time.in.sec, &#39; seconds&#39;, &#39;\n&#39;);</span>
<span class="p">}</span><span class="n">else</span><span class="p">{</span>
	<span class="n">output_file_name</span> <span class="o">=</span> <span class="nf">tempfile</span><span class="p">();</span>
	<span class="n">meq</span> <span class="o">=</span> <span class="nf">Matrix_eQTL_engine</span><span class="p">(</span><span class="n">snps</span> <span class="o">=</span> <span class="n">snps</span><span class="p">,</span><span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span><span class="n">cvrt</span> <span class="o">=</span> <span class="n">cvrt</span><span class="p">,</span>
		<span class="n">output_file_name</span> <span class="o">=</span> <span class="n">output_file_name</span><span class="p">,</span><span class="n">pvOutputThreshold</span> <span class="o">=</span> <span class="n">pvOutputThreshold</span><span class="p">,</span>
		<span class="n">useModel</span> <span class="o">=</span> <span class="n">useModel</span><span class="p">,</span>
		<span class="c1"># errorCovariance = errorCovariance,</span>
		<span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
		<span class="n">pvalue.hist</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span><span class="n">min.pv.by.genesnp</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span><span class="n">noFDRsaveMemory</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">);</span>
	<span class="n">meh</span> <span class="o">=</span> <span class="nf">Matrix_eQTL_engine</span><span class="p">(</span>
		<span class="n">snps</span> <span class="o">=</span> <span class="n">snps</span><span class="p">,</span>
		<span class="n">gene</span> <span class="o">=</span> <span class="n">gene</span><span class="p">,</span>
		<span class="n">cvrt</span> <span class="o">=</span> <span class="n">cvrt</span><span class="p">,</span>
		<span class="n">output_file_name</span> <span class="o">=</span> <span class="n">output_file_name</span><span class="p">,</span>
		<span class="n">pvOutputThreshold</span> <span class="o">=</span> <span class="n">pvOutputThreshold</span><span class="p">,</span>
		<span class="n">useModel</span> <span class="o">=</span> <span class="n">useModel</span><span class="p">,</span>
		<span class="c1"># errorCovariance = numeric(),</span>
		<span class="n">verbose</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span>
		<span class="n">pvalue.hist</span> <span class="o">=</span> <span class="m">100</span><span class="p">);</span>
	<span class="nf">unlink</span><span class="p">(</span> <span class="n">filename</span> <span class="p">);</span>
	<span class="nf">unlink</span><span class="p">(</span><span class="n">output_file_name</span><span class="p">);</span>
	<span class="nf">write.csv</span><span class="p">(</span><span class="n">meq</span><span class="o">$</span><span class="n">all</span><span class="o">$</span><span class="n">eqtls</span><span class="p">,</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#39;eqtl.csv&#39;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">),</span><span class="n">row.names</span><span class="o">=</span><span class="bp">F</span><span class="p">,</span><span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
<span class="c1"># cat(&#39;Analysis done in: &#39;, meq$time.in.sec+meh$time.in.sec, &#39; seconds&#39;, &#39;\n&#39;);</span>
<span class="p">}</span>
<span class="nf">save.image</span><span class="p">(</span><span class="s">&#34;calculat.RData&#34;</span><span class="p">)</span>
<span class="c1">## Plot the Q-Q plot </span>
<span class="nf">png</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;Q-Q.png&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">meq</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="nf">pdf</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;Q-Q.pdf&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">meq</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1">## Plot the histogram plot</span>
<span class="nf">png</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;histogram.png&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">meh</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s">&#34;grey&#34;</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="c1">## Plot the histogram plot</span>
<span class="nf">pdf</span><span class="p">(</span><span class="nf">paste</span><span class="p">(</span><span class="n">opt</span><span class="o">$</span><span class="n">outdir</span><span class="p">,</span><span class="s">&#34;histogram.pdf&#34;</span><span class="p">,</span><span class="n">sep</span><span class="o">=</span><span class="s">&#39;&#39;</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">meh</span><span class="p">,</span><span class="n">col</span><span class="o">=</span><span class="s">&#34;grey&#34;</span><span class="p">)</span>
<span class="nf">dev.off</span><span class="p">()</span>

<span class="nf">save.image</span><span class="p">(</span><span class="s">&#34;eQTL.RData&#34;</span><span class="p">)</span>
</code></pre></td></tr></table>
</div>
</div><h3 id="结果展示-2">结果展示：</h3>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step03.result.png"
        data-srcset="/posts/matrix-eqtl-twas/step03.result.png, step03.result.png 1.5x, /posts/matrix-eqtl-twas/step03.result.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step03.result.png"
        title="step03.result.png" /></p>
<p><img
        class="lazyload"
        src="/svg/loading.min.svg"
        data-src="step03.result-details.png"
        data-srcset="/posts/matrix-eqtl-twas/step03.result-details.png, step03.result-details.png 1.5x, /posts/matrix-eqtl-twas/step03.result-details.png 2x"
        data-sizes="auto"
        alt="/posts/matrix-eqtl-twas/step03.result-details.png"
        title="step03.result-details.png" /></p>
<h2 id="参考文献">参考文献</h2>
<p>Andrey A. Shabalin, <a href="https://doi.org/10.1093/bioinformatics/bts163" target="_blank" rel="noopener noreffer">Matrix eQTL: ultra fast eQTL analysis via large matrix operations</a>, Bioinformatics, Volume 28, Issue 10, 15 May 2012, Pages 1353–1358.</p>
</div><div class="post-footer" id="post-footer">
    <div class="post-info">
        <div class="post-info-line">
            <div class="post-info-mod">
                <span>更新于 2020-12-18</span>
            </div>
            <div class="post-info-license"></div>
        </div>
        <div class="post-info-line">
            <div class="post-info-md"><span>
                            <a class="link-to-markdown" href="/posts/matrix-eqtl-twas/index.md" target="_blank">阅读原始文档</a>
                        </span></div>
            <div class="post-info-share">
                <span><a href="javascript:void(0);" title="分享到 Twitter" data-sharer="twitter" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" data-title="TWAS/eQTL analysis pipeline" data-via="https://twitter.com/czheluo" data-hashtags="eQTL,TWAS,RNAseq,表达量,定位"><i class="fab fa-twitter fa-fw"></i></a><a href="javascript:void(0);" title="分享到 Facebook" data-sharer="facebook" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" data-hashtag="eQTL"><i class="fab fa-facebook-square fa-fw"></i></a><a href="javascript:void(0);" title="分享到 Linkedin" data-sharer="linkedin" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/"><i class="fab fa-linkedin fa-fw"></i></a><a href="javascript:void(0);" title="分享到 Hacker News" data-sharer="hackernews" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" data-title="TWAS/eQTL analysis pipeline"><i class="fab fa-hacker-news fa-fw"></i></a><a href="javascript:void(0);" title="分享到 Line" data-sharer="line" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" data-title="TWAS/eQTL analysis pipeline"><i data-svg-src="/lib/simple-icons/icons/line.min.svg"></i></a><a href="javascript:void(0);" title="分享到 微博" data-sharer="weibo" data-url="http://mojorbioinfor.work/posts/matrix-eqtl-twas/" data-title="TWAS/eQTL analysis pipeline" data-ralateuid="https://weibo.com/2161302013/profile"><i class="fab fa-weibo fa-fw"></i></a></span>
            </div>
        </div>
    </div>

    <div class="post-info-more">
        <section class="post-tags"><i class="fas fa-tags fa-fw"></i>&nbsp;<a href="/tags/eqtl/">eQTL</a>,&nbsp;<a href="/tags/twas/">TWAS</a>,&nbsp;<a href="/tags/rnaseq/">RNAseq</a>,&nbsp;<a href="/tags/%E8%A1%A8%E8%BE%BE%E9%87%8F/">表达量</a>,&nbsp;<a href="/tags/%E5%AE%9A%E4%BD%8D/">定位</a></section>
        <section>
            <span><a href="javascript:void(0);" onclick="window.history.back();">返回</a></span>&nbsp;|&nbsp;<span><a href="/">主页</a></span>
        </section>
    </div>

    <div class="post-nav"><a href="/posts/probabilistic-estimation-of-expression-residuals/" class="prev" rel="prev" title="Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses"><i class="fas fa-angle-left fa-fw"></i>Using probabilistic estimation of expression residuals (PEER) to obtain increased power and interpretability of gene expression analyses</a>
            <a href="/posts/stem-sop/" class="next" rel="next" title="The Short Time-series Expression Miner">The Short Time-series Expression Miner<i class="fas fa-angle-right fa-fw"></i></a></div>
</div>
<div id="comments"><div id="valine" class="comment"></div><noscript>
                Please enable JavaScript to view the comments powered by <a href="https://valine.js.org/">Valine</a>.
            </noscript><div id="fb-root" class="comment"></div>
            <div
                class="fb-comments"
                data-href="http://mojorbioinfor.work/posts/matrix-eqtl-twas/"
                data-width="100%"
                data-numposts="10"
            ></div><noscript>
                Please enable JavaScript to view the comments powered by <a href="https://developers.facebook.com/docs/plugins/comments/"></a>Facebook</a>.
            </noscript></div></article></div>
            </main><footer class="footer">
        <div class="footer-container"><div class="footer-line">MajorBioInformatics</div><div class="footer-line">Powered By <a href="https://gohugo.io/" target="_blank" rel="noopener noreffer" title="Hugo 0.79.0">Hugo</a> | Theme - <a href="https://github.com/dillonzq/LoveIt" target="_blank" rel="noopener noreffer" title="LoveIt 0.2.10"> LoveIt</a> 
                    
                </div>
                <div class="footer-line"><i class="far fa-copyright fa-fw"></i><span itemprop="copyrightYear">2020</span><span class="author" itemprop="copyrightHolder">&nbsp;<a href="/" target="_blank"></a></span>&nbsp;|&nbsp;<span class="license"><a rel="license external nofollow noopener noreffer" href="https://creativecommons.org/licenses/by-nc/4.0/" target="_blank">CC BY-NC 4.0</a></span></div>
        </div>
    </footer></div>

        <div id="fixed-buttons"><a href="#" id="back-to-top" class="fixed-button" title="回到顶部">
                <i class="fas fa-arrow-up fa-fw"></i>
            </a><a href="#" id="view-comments" class="fixed-button" title="查看评论">
                <i class="fas fa-comment fa-fw"></i>
            </a>
        </div><link rel="stylesheet" href="/lib/valine/valine.min.css"><link rel="stylesheet" href="/lib/lightgallery/lightgallery.min.css"><link rel="stylesheet" href="/lib/katex/katex.min.css"><link rel="stylesheet" href="/lib/katex/copy-tex.min.css"><script type="text/javascript" src="/lib/valine/Valine.min.js"></script><script type="text/javascript" src="https://connect.facebook.net/zh_CN/sdk.js#xfbml=1&amp;version=v5.0&amp;appId=&amp;autoLogAppEvents=1" defer></script><script type="text/javascript" src="/lib/smooth-scroll/smooth-scroll.min.js"></script><script type="text/javascript" src="/lib/autocomplete/autocomplete.min.js"></script><script type="text/javascript" src="/lib/algoliasearch/algoliasearch-lite.umd.min.js"></script><script type="text/javascript" src="/lib/lazysizes/lazysizes.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lightgallery.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-thumbnail.min.js"></script><script type="text/javascript" src="/lib/lightgallery/lg-zoom.min.js"></script><script type="text/javascript" src="/lib/clipboard/clipboard.min.js"></script><script type="text/javascript" src="/lib/sharer/sharer.min.js"></script><script type="text/javascript" src="/lib/katex/katex.min.js"></script><script type="text/javascript" src="/lib/katex/auto-render.min.js"></script><script type="text/javascript" src="/lib/katex/copy-tex.min.js"></script><script type="text/javascript" src="/lib/katex/mhchem.min.js"></script><script type="text/javascript">window.config={"code":{"copyTitle":"复制到剪贴板","maxShownLines":10},"comment":{"valine":{"appId":"Noyy9UKaXGwYBIUdvBTt7u76-gzGzoHsz","appKey":"15w6pTF8mFoxnm8muOFx5buM","avatar":"mp","el":"#valine","emojiCDN":"https://cdn.jsdelivr.net/npm/emoji-datasource-google@5.0.1/img/google/64/","emojiMaps":{"100":"1f4af.png","alien":"1f47d.png","anger":"1f4a2.png","angry":"1f620.png","anguished":"1f627.png","astonished":"1f632.png","black_heart":"1f5a4.png","blue_heart":"1f499.png","blush":"1f60a.png","bomb":"1f4a3.png","boom":"1f4a5.png","broken_heart":"1f494.png","brown_heart":"1f90e.png","clown_face":"1f921.png","cold_face":"1f976.png","cold_sweat":"1f630.png","confounded":"1f616.png","confused":"1f615.png","cry":"1f622.png","crying_cat_face":"1f63f.png","cupid":"1f498.png","dash":"1f4a8.png","disappointed":"1f61e.png","disappointed_relieved":"1f625.png","dizzy":"1f4ab.png","dizzy_face":"1f635.png","drooling_face":"1f924.png","exploding_head":"1f92f.png","expressionless":"1f611.png","face_vomiting":"1f92e.png","face_with_cowboy_hat":"1f920.png","face_with_hand_over_mouth":"1f92d.png","face_with_head_bandage":"1f915.png","face_with_monocle":"1f9d0.png","face_with_raised_eyebrow":"1f928.png","face_with_rolling_eyes":"1f644.png","face_with_symbols_on_mouth":"1f92c.png","face_with_thermometer":"1f912.png","fearful":"1f628.png","flushed":"1f633.png","frowning":"1f626.png","ghost":"1f47b.png","gift_heart":"1f49d.png","green_heart":"1f49a.png","grimacing":"1f62c.png","grin":"1f601.png","grinning":"1f600.png","hankey":"1f4a9.png","hear_no_evil":"1f649.png","heart":"2764-fe0f.png","heart_decoration":"1f49f.png","heart_eyes":"1f60d.png","heart_eyes_cat":"1f63b.png","heartbeat":"1f493.png","heartpulse":"1f497.png","heavy_heart_exclamation_mark_ornament":"2763-fe0f.png","hole":"1f573-fe0f.png","hot_face":"1f975.png","hugging_face":"1f917.png","hushed":"1f62f.png","imp":"1f47f.png","innocent":"1f607.png","japanese_goblin":"1f47a.png","japanese_ogre":"1f479.png","joy":"1f602.png","joy_cat":"1f639.png","kiss":"1f48b.png","kissing":"1f617.png","kissing_cat":"1f63d.png","kissing_closed_eyes":"1f61a.png","kissing_heart":"1f618.png","kissing_smiling_eyes":"1f619.png","laughing":"1f606.png","left_speech_bubble":"1f5e8-fe0f.png","love_letter":"1f48c.png","lying_face":"1f925.png","mask":"1f637.png","money_mouth_face":"1f911.png","nauseated_face":"1f922.png","nerd_face":"1f913.png","neutral_face":"1f610.png","no_mouth":"1f636.png","open_mouth":"1f62e.png","orange_heart":"1f9e1.png","partying_face":"1f973.png","pensive":"1f614.png","persevere":"1f623.png","pleading_face":"1f97a.png","pouting_cat":"1f63e.png","purple_heart":"1f49c.png","rage":"1f621.png","relaxed":"263a-fe0f.png","relieved":"1f60c.png","revolving_hearts":"1f49e.png","right_anger_bubble":"1f5ef-fe0f.png","robot_face":"1f916.png","rolling_on_the_floor_laughing":"1f923.png","scream":"1f631.png","scream_cat":"1f640.png","see_no_evil":"1f648.png","shushing_face":"1f92b.png","skull":"1f480.png","skull_and_crossbones":"2620-fe0f.png","sleeping":"1f634.png","sleepy":"1f62a.png","slightly_frowning_face":"1f641.png","slightly_smiling_face":"1f642.png","smile":"1f604.png","smile_cat":"1f638.png","smiley":"1f603.png","smiley_cat":"1f63a.png","smiling_face_with_3_hearts":"1f970.png","smiling_imp":"1f608.png","smirk":"1f60f.png","smirk_cat":"1f63c.png","sneezing_face":"1f927.png","sob":"1f62d.png","space_invader":"1f47e.png","sparkling_heart":"1f496.png","speak_no_evil":"1f64a.png","speech_balloon":"1f4ac.png","star-struck":"1f929.png","stuck_out_tongue":"1f61b.png","stuck_out_tongue_closed_eyes":"1f61d.png","stuck_out_tongue_winking_eye":"1f61c.png","sunglasses":"1f60e.png","sweat":"1f613.png","sweat_drops":"1f4a6.png","sweat_smile":"1f605.png","thinking_face":"1f914.png","thought_balloon":"1f4ad.png","tired_face":"1f62b.png","triumph":"1f624.png","two_hearts":"1f495.png","unamused":"1f612.png","upside_down_face":"1f643.png","weary":"1f629.png","white_frowning_face":"2639-fe0f.png","white_heart":"1f90d.png","wink":"1f609.png","woozy_face":"1f974.png","worried":"1f61f.png","yawning_face":"1f971.png","yellow_heart":"1f49b.png","yum":"1f60b.png","zany_face":"1f92a.png","zipper_mouth_face":"1f910.png","zzz":"1f4a4.png"},"enableQQ":false,"highlight":true,"lang":"zh-cn","pageSize":10,"placeholder":"你的评论 ...","recordIP":true,"visitor":true}},"lightGallery":{"actualSize":false,"exThumbImage":"data-thumbnail","hideBarsDelay":2000,"selector":".lightgallery","speed":400,"thumbContHeight":80,"thumbWidth":80,"thumbnail":true},"math":{"delimiters":[{"display":true,"left":"$$","right":"$$"},{"display":true,"left":"\\[","right":"\\]"},{"display":false,"left":"$","right":"$"},{"display":false,"left":"\\(","right":"\\)"}],"strict":false},"search":{"algoliaAppID":"VX4BXOJBC5","algoliaIndex":"majorbio.zh-cn","algoliaSearchKey":"53790f167edb548b217bfe91dd891ac0","highlightTag":"em","maxResultLength":10,"noResultsFound":"没有找到结果","snippetLength":50,"type":"algolia"}};</script><script type="text/javascript" src="/js/theme.min.js"></script></body>
</html>
